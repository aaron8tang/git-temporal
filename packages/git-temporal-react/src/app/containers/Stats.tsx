import React, { Component } from 'react';
import { connect } from 'react-redux';

import { DispatchProps } from 'app/interfaces';
import { getFilteredStats } from 'app/selectors';
import { style } from 'app/styles';
import { StackedLabelHeader } from '../components/StackedLabelHeader';
import { CommaNumber } from '../components/CommaNumber';

interface StatsProps {
  minAuthorDate?: number;
  maxAuthorDate?: number;
  authors?: number;
  commits?: number;
  files?: number;
  linesAdded?: number;
  linesDeleted?: number;
}

export class Stats extends Component<StatsProps & DispatchProps> {
  render() {
    return (
      <div style={style('panel', 'flexRows')}>
        <StackedLabelHeader label="Contributors">
          <CommaNumber value={this.props.authors.toString()} />
        </StackedLabelHeader>
        <StackedLabelHeader label="Active Time Span">
          {`${this.props.minAuthorDate} - ${this.props.maxAuthorDate}`}
        </StackedLabelHeader>
        <StackedLabelHeader label="Lines Added">
          <CommaNumber value={this.props.linesAdded.toString()} />
        </StackedLabelHeader>
        <StackedLabelHeader label="Lines Deleted">
          <CommaNumber value={this.props.linesDeleted.toString()} />
        </StackedLabelHeader>
        <StackedLabelHeader label="Commits">
          <CommaNumber value={this.props.commits.toString()} />
        </StackedLabelHeader>
        <StackedLabelHeader label="Files">
          <CommaNumber value={this.props.files.toString()} />
        </StackedLabelHeader>
      </div>
    );
  }
}

export const mapStateToProps = state => {
  return getFilteredStats(state);
};

export default connect(mapStateToProps)(Stats);
