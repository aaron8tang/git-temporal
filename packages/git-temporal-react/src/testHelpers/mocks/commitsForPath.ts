import fiveCommits from './fiveCommits';

const commitsForPath = {
  selectedPath: 'testPath1',
  isFetching: false,
  didInvalidate: false,

  commits: fiveCommits,
};

export default commitsForPath;
