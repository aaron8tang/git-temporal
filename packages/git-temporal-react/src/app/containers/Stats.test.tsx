import React from 'react';
import { shallow } from 'enzyme';

import { Stats, mapStateToProps, CommitsOrFiles } from './Stats';
import commitsForPath from 'testHelpers/mocks/commitsForPath';

const testProps = {
  authors: 5,
  minAuthorDate: 1536691136,
  maxAuthorDate: 1538663085,
  commits: 52,
  files: 110,
  linesAdded: 700,
  linesDeleted: 300,
  isFileSelected: false,
  viewCommitsOrFiles: CommitsOrFiles.COMMITS,
};

global.Date.now = jest.fn(() => 1539563458 * 1000);

describe('containers/Stats', () => {
  describe('when rendered test props', () => {
    let wrapper;
    let mockDispatch;
    beforeAll(() => {
      mockDispatch = jest.fn();
      wrapper = shallow(<Stats dispatch={mockDispatch} {...testProps} />);
    });
    beforeEach(() => {
      mockDispatch.mockClear();
    });

    test('it should match snapshot (it should be showing empty indicator)', () => {
      expect(wrapper).toMatchSnapshot();
    });
    test('when clicking on commits it should call dispatch', () => {
      wrapper.instance().onCommitsClick();
      expect(mockDispatch).toHaveBeenCalledTimes(1);
    });
    test('when clicking on files it should call dispatch', () => {
      wrapper.instance().onFilesClick();
      expect(mockDispatch).toHaveBeenCalledTimes(1);
    });
  });
  describe('when rendered and a directory is selected', () => {
    let wrapper;
    let mockDispatch;
    beforeAll(() => {
      const dirTestProps = Object.assign({}, testProps, {
        isFileSelected: true,
        viewCommitsOrFiles: CommitsOrFiles.FILES,
      });
      mockDispatch = jest.fn();
      wrapper = shallow(<Stats dispatch={mockDispatch} {...dirTestProps} />);
    });
    beforeEach(() => {
      mockDispatch.mockClear();
    });

    test('it should match snapshot (it should be showing Commits selected)', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('it should call dispatch to change to CommitsOrFiles.COMMITS onComponentDidMount', () => {
      wrapper.instance().componentDidMount();
      expect(mockDispatch).toHaveBeenCalledTimes(1);
    });

    test('it should call dispatch to change to CommitsOrFiles.COMMITS onComponentDidUpdate', () => {
      wrapper.instance().componentDidUpdate();
      expect(mockDispatch).toHaveBeenCalledTimes(1);
    });
  });

  describe('when calling mapStateToProps()', () => {
    test('it should respond with props', () => {
      const propsOut = mapStateToProps(commitsForPath);
      expect(propsOut).toMatchSnapshot();
    });
  });
});
