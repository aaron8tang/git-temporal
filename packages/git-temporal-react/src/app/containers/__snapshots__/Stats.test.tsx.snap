// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Stats when calling mapStateToProps() it should respond with props 1`] = `
Object {
  "authors": 5,
  "commits": 5,
  "files": 13,
  "isFileSelected": false,
  "linesAdded": 96,
  "linesDeleted": 667,
  "maxAuthorDate": 1537541919,
  "minAuthorDate": 1526296167,
  "viewCommitsOrFiles": "commits",
}
`;

exports[`containers/Stats when rendered and a directory is selected it should match snapshot (it should be showing Commits selected) 1`] = `
<div
  style={
    Object {
      "backgroundColor": "white",
      "borderRadius": 3,
      "display": "flex",
      "flexDirection": "row",
      "marginRight": 10,
      "padding": 10,
    }
  }
>
  <Component
    label="Contributors"
  >
    <Component
      value={5}
    />
  </Component>
  <Component
    label="Active Time Span"
  >
    <Component
      firstEpochTime={1536691136}
      secondEpochTime={1538663085}
    />
  </Component>
  <Component
    label="Last Commit"
  >
    <div>
      <Component
        firstEpochTime={1538663085}
        secondEpochTime={1539563458}
      />
      <span>
         ago
      </span>
    </div>
  </Component>
  <Component
    label="Lines Added"
  >
    <Component
      value={700}
    />
  </Component>
  <Component
    label="Lines Deleted"
  >
    <Component
      value={300}
    />
  </Component>
  <Component
    isSelected={false}
    label="Commits"
    onLabelClick={[Function]}
    title="Click to view Commits"
  >
    <Component
      value={52}
    />
  </Component>
  <Component
    isSelected={true}
    label="Files"
    onLabelClick={null}
    title="Click to view Files"
  >
    <Component
      value={110}
    />
  </Component>
</div>
`;

exports[`containers/Stats when rendered test props it should match snapshot (it should be showing empty indicator) 1`] = `
<div
  style={
    Object {
      "backgroundColor": "white",
      "borderRadius": 3,
      "display": "flex",
      "flexDirection": "row",
      "marginRight": 10,
      "padding": 10,
    }
  }
>
  <Component
    label="Contributors"
  >
    <Component
      value={5}
    />
  </Component>
  <Component
    label="Active Time Span"
  >
    <Component
      firstEpochTime={1536691136}
      secondEpochTime={1538663085}
    />
  </Component>
  <Component
    label="Last Commit"
  >
    <div>
      <Component
        firstEpochTime={1538663085}
        secondEpochTime={1539563458}
      />
      <span>
         ago
      </span>
    </div>
  </Component>
  <Component
    label="Lines Added"
  >
    <Component
      value={700}
    />
  </Component>
  <Component
    label="Lines Deleted"
  >
    <Component
      value={300}
    />
  </Component>
  <Component
    isSelected={true}
    label="Commits"
    onLabelClick={[Function]}
    title="Click to view Commits"
  >
    <Component
      value={52}
    />
  </Component>
  <Component
    isSelected={false}
    label="Files"
    onLabelClick={[Function]}
    title="Click to view Files"
  >
    <Component
      value={110}
    />
  </Component>
</div>
`;
